diff --git a/app/src/main/java/com/bugsnag/android/DefaultDelivery.kt b/app/src/main/java/com/bugsnag/android/DefaultDelivery.kt
index 0ce2eec8c4..e1bac196e2 100644
--- a/app/src/main/java/com/bugsnag/android/DefaultDelivery.kt
+++ b/app/src/main/java/com/bugsnag/android/DefaultDelivery.kt
@@ -66,7 +66,7 @@ internal class DefaultDelivery(
             return DeliveryStatus.UNDELIVERED
         } catch (exception: IOException) {
             logger.w("IOException encountered in request", exception)
-            return DeliveryStatus.UNDELIVERED
+            return DeliveryStatus.FAILURE
         } catch (exception: Exception) {
             logger.w("Unexpected error delivering payload", exception)
             return DeliveryStatus.FAILURE
