diff --git a/app/src/main/java/androidx/emoji2/text/SpannableBuilder.java b/app/src/main/java/androidx/emoji2/text/SpannableBuilder.java
index 366c7b3629..e6fef5cddb 100644
--- a/app/src/main/java/androidx/emoji2/text/SpannableBuilder.java
+++ b/app/src/main/java/androidx/emoji2/text/SpannableBuilder.java
@@ -46,7 +46,7 @@ import java.util.concurrent.atomic.AtomicInteger;
  *
  */
 @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
-public final class SpannableBuilder extends SpannableStringBuilder {
+public final class SpannableBuilder extends eu.faircode.email.SpannableStringBuilderEx {
     /**
      * DynamicLayout$ChangeWatcher class.
      */
