diff --git a/app/src/main/java/eu/faircode/email/HtmlEx.java b/app/src/main/java/eu/faircode/email/HtmlEx.java
index c8b8e1b8f..5cfceaccd 100644
--- a/app/src/main/java/eu/faircode/email/HtmlEx.java
+++ b/app/src/main/java/eu/faircode/email/HtmlEx.java
@@ -276,7 +276,7 @@ public class HtmlEx {
 
     private /* static */ void withinBlockquoteConsecutive(StringBuilder out, Spanned text, int start,
                                                     int end) {
-        out.append("<p").append(getTextDirection(text, start, end)).append(">");
+        out.append("<span").append(getTextDirection(text, start, end)).append(">");
 
         int next;
         for (int i = start; i < end; i = next) {
@@ -294,21 +294,21 @@ public class HtmlEx {
 
             withinParagraph(out, text, i, next - nl);
 
-            if (nl == 1) {
+            if (nl == 0) {
                 out.append("<br>\n");
             } else {
-                for (int j = 2; j < nl; j++) {
+                for (int j = 0; j < nl; j++) {
                     out.append("<br>");
                 }
                 if (next != end) {
                     /* Paragraph should be closed and reopened */
-                    out.append("</p>\n");
-                    out.append("<p").append(getTextDirection(text, start, end)).append(">");
+                    out.append("</span>\n");
+                    out.append("<span").append(getTextDirection(text, start, end)).append(">");
                 }
             }
         }
 
-        out.append("</p>\n");
+        out.append("</span>\n");
     }
 
     private /* static */ void withinParagraph(StringBuilder out, Spanned text, int start, int end) {
