diff --git a/crowdin.py b/crowdin.py
old mode 100755
new mode 100644
diff --git a/helper.py b/helper.py
index d3a57c9..7ef7d14 100644
--- a/helper.py
+++ b/helper.py
@@ -109,8 +109,12 @@ def GetResDirectory(path):
 
 
 def GetTargetStringsXml(targetResDirectory, langCode):
-    country = langCode[0:2]
-    variant = langCode[3:]
+    if langCode == 'fil':
+      country = 'fil'
+      variant = ''
+    else:
+      country = langCode[0:2]
+      variant = langCode[3:]
     if variant is not None and len(variant) > 0:
         variant = "-r" + variant
 
@@ -127,4 +131,4 @@ def IsSingleFolderUpdate(path):
 def IsDefaultStringsXml(path):
     if "res/values/strings.xml" in path:
         return True
-    return False
\ No newline at end of file
+    return False
